apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/MetricRegistry.java

```diff
@@ -40,93 +45,196 @@
      * Adds a gauge to the metric registry.
      *
      * @param name   the name of the metric
-     * @param tags   tags for the metric.
+     * @param labels labels for the metric.
      *               Tags can be used to create different graphs based for each value of a specific tag name, using a terms aggregation.
-     *               Note that there will be a {@link MetricSet} created for each distinct set of tags.
+     *               Note that there will be a {@link MetricSet} created for each distinct set of labels.
      * @param metric this supplier will be called for every reporting cycle
      *               ({@link co.elastic.apm.agent.report.ReporterConfiguration#metricsInterval metrics_interval)})
      */
-    public void add(String name, Map<String, String> tags, DoubleSupplier metric) {
+    public void add(String name, Labels labels, DoubleSupplier metric) {
         if (isDisabled(name)) {
             return;
         }
-        MetricSet metricSet = metricSets.get(tags);
-        if (metricSet == null) {
-            metricSets.putIfAbsent(tags, new MetricSet(tags));
-            metricSet = metricSets.get(tags);
+
+        long criticalValueAtEnter = phaser.writerCriticalSectionEnter();
+        try {
+            final MetricSet metricSet = getOrCreateMetricSet(labels);
+            if (metricSet != null) {
+                metricSet.addGauge(name, metric);
+            }
+        } finally {
+            phaser.writerCriticalSectionExit(criticalValueAtEnter);
         }
-        metricSet.add(name, metric);
     }
 
     private boolean isDisabled(String name) {
         return WildcardMatcher.anyMatch(config.getDisableMetrics(), name) != null;
     }
 
-    public double get(String name, Map<String, String> tags) {
-        final MetricSet metricSet = metricSets.get(tags);
+    public double getGauge(String name, Labels labels) {
+        final MetricSet metricSet = activeMetricSets.get(labels);
         if (metricSet != null) {
-            return metricSet.get(name).get();
+            return metricSet.getGauge(name).get();
         }
         return Double.NaN;
     }
 
-    public Map<Map<String, String>, MetricSet> getMetricSets() {
-        return metricSets;
+    public void report(MetricsReporter metricsReporter) {
+        try {
+            phaser.readerLock();
+            ConcurrentMap<Labels.Immutable, MetricSet> temp = inactiveMetricSets;
+            inactiveMetricSets = activeMetricSets;
+            activeMetricSets = temp;
+            phaser.flipPhase();
+            metricsReporter.report(inactiveMetricSets);
+        } finally {
+            phaser.readerUnlock();
+        }
+    }
+
+    public void updateTimer(String timerName, Labels labels, long durationUs) {
+        updateTimer(timerName, labels, durationUs, 1);
+    }
+
+    public void updateTimer(String timerName, Labels labels, long durationUs, long count) {
+        long criticalValueAtEnter = phaser.writerCriticalSectionEnter();
```

Comment:
I think in the current scenario this is unnecessary, as it is guarded from the outside. Maybe that's the way to go - pass the guarding responsibility to the MethodRegistry user 🤔